### PR TITLE
add handler for sigterm and use exec to run server.py script

### PR DESCRIPTION
**Pull request by jsanda**
This PR includes a couple changes:

* Adds a signal handler for SIGTERM
    * When invoked the handler logs that the server is shutting down
    * The handler calls `server.stop()` on the gRPC serve
* Use `exec` to run `server.py` so that it is not run in a subshell

### CODE REVIEW

File: medusa/service/grpc/server.py
Comment (adejanovski) [inline medusa/service/grpc/server.py:176]: Would it be possible to remove these lines if they're not used anymore?

File: k8s/docker-entrypoint.sh
Comment (adejanovski) [inline k8s/docker-entrypoint.sh:51]: Out of curiosity, what's the difference with using `exec` instead of calling `python3` directly? Is it related to signal handling?

**Review by adejanovski — CHANGES_REQUESTED** [review]
There's a comment that looks like it could be removed and maybe it could be nice to have a small test somewhere that starts the server and then sends a SIGTERM to verify that the change works while protecting us from future regression.
wdyt?

File: medusa/service/grpc/server.py
Comment (jsanda) [reply medusa/service/grpc/server.py:176]: >There's a comment that looks like it could be removed and maybe it could be nice to have a small test somewhere that starts the server and then sends a SIGTERM to verify that the change works while protecting us from future regression.
wdyt?

Definitely. I will work on a test.

>Would it be possible to remove these lines if they're not used anymore?

👍

File: k8s/docker-entrypoint.sh
Comment (jsanda) [reply k8s/docker-entrypoint.sh:51]: `exec` causes the new process to be run with in the current shell/process.

Here is the `ps` output from within a `medusa` container that does **not** use `exec`:

```
$ ps -x
    PID TTY      STAT   TIME COMMAND
      1 ?        Ss     0:00 /bin/bash /home/cassandra/docker-entrypoint.sh
      7 ?        Sl     0:00 python3 -m medusa.service.grpc.server server.py
    372 pts/0    Ss     0:00 /bin/bash
```
Running `kill -SIGTERM 1` has no effect here. 

Here is the `ps` output from a `medusa` container where `exec` is used:

```
$ ps -x
    PID TTY      STAT   TIME COMMAND
      1 ?        Ssl    0:00 python3 -m medusa.service.grpc.server server.py
    347 pts/0    Ss     0:00 /bin/bash
```
Running `kill -SIGTERM 1` has the desired effect here.

File: k8s/docker-entrypoint.sh
Comment (emerkle826) [inline k8s/docker-entrypoint.sh:61]: If changing the `grpc()` function above in this script has the desired affect, can we not simply call the function here (like what was done before)?

Also, would we need to invoke `exec` on line 44 [here](https://github.com/thelastpickle/cassandra-medusa/pull/313/files#diff-7e552c6093dd30777eac14ab0ab705b391b0d963d156ccda72178fef2b3514a6R44) as well, or is there no issue for restore?

**Review by emerkle826 — CHANGES_REQUESTED** [review]
This looks ok to me. Just a couple questions about `exec` and restore that maybe do not need to be addressed.

File: k8s/docker-entrypoint.sh
Comment (jsanda) [reply k8s/docker-entrypoint.sh:61]: I need to clean this up. Provided the function call does not result in a new shell/process we should be able to leave it as is and just do the `exec` inside the `grpc` function. I will test.

>Also, would we need to invoke exec on line 44 here as well, or is there no issue for restore?

Initially I didn't think we needed it for restore since it is a long-live server process; however a restore operation could take a long time. I think we probably want it for restore as well to be on the safe side.

File: medusa/service/grpc/server.py
Comment (jsanda) [reply medusa/service/grpc/server.py:176]: With respect to testing, the issue is specific to containerized environments. If you run the gRPC server locally and send it `SIGTERM`, the python process will terminate. The integration tests already performa shutdown on the gRPC server, but they do not run in a containerized environment

Comment (adejanovski) [reply medusa/service/grpc/server.py]: gotcha 👍

File: k8s/docker-entrypoint.sh
Comment (adejanovski) [reply k8s/docker-entrypoint.sh:51]: kk, thanks

**Comment by jsanda** [comment]
@adejanovski can I get another review on this?

**Review by adejanovski — APPROVED** [review]
LGTM 👍 
Let's wait for CI to turn green and then you can squash the commits and merge the PR.

**Pull request merged**
